src/libostree/ostree-async-progress.c

```diff
@@ -157,6 +157,68 @@ ostree_async_progress_get_uint64 (OstreeAsyncProgress       *self,
   return (rval != NULL) ? g_variant_get_uint64 (rval) : 0;
 }
 
+/**
+ * ostree_async_progress_get:
+ * @self: an #OstreeAsyncProgress
+ * @...: key name, format string, #GVariant return locations, …, followed by %NULL
+ *
+ * Get the values corresponding to zero or more keys from the
+ * #OstreeAsyncProgress. Each key is specified in @... as the key name, followed
+ * by a #GVariant format string, followed by the necessary arguments for that
+ * format string, just as for g_variant_get(). After those arguments is the
+ * next key name. The varargs list must be %NULL-terminated.
+ *
+ * Each format string must make deep copies of its value, as the values stored
+ * in the #OstreeAsyncProgress may be freed from another thread after this
+ * function returns.
+ *
+ * This operation is thread-safe, and all the keys are queried atomically.
+ *
+ * |[<!-- language="C" -->
+ * guint32 outstanding_fetches;
+ * guint64 bytes_received;
+ * g_autofree gchar *status = NULL;
+ * g_autoptr(GVariant) refs_variant = NULL;
+ *
+ * ostree_async_progress_get (progress,
+ *                            "outstanding-fetches", "u", &outstanding_fetches,
+ *                            "bytes-received", "t", &bytes_received,
+ *                            "status", "s", &status,
+ *                            "refs", "@a{ss}", &refs_variant,
+ *                            NULL);
+ * ]|
+ *
+ * Since: 2017.6
+ */
+void
+ostree_async_progress_get (OstreeAsyncProgress *self,
+                           ...)
+{
+  va_list ap;
+  const char *key, *format_string;
+
+  g_mutex_lock (&self->lock);
+  va_start (ap, self);
+
+  for (key = va_arg (ap, const char *), format_string = va_arg (ap, const char *);
+       key != NULL;
+       key = va_arg (ap, const char *), format_string = va_arg (ap, const char *))
```

Comment:
IMO this would be more readable if this was:
```
while ((key = va_arg (ap, const char *)) != NULL) {
    format_string = va_arg (ap, const char *);
    ...
}
```